Encryption.md

```diff
@@ -19,216 +19,544 @@
+"The random field shall .. consist of an output string of r(i) bits from an approved RBG with a 
+sufficient security strength". That is, a fresh random byte array must be generated for each nonce - 
+as opposed to the option 2 in the same spec - "or the result of applying the r(i)–bit incrementing 
+function to the random field of the preceding <nonce> for the given key"
+
+
+### 4.2 AES_GCM_CTR_V1 encryption algorithm
+In this Parquet algorithm, all Thrift modules are encrypted with the GCM cipher, as described above. 
+
+The pages are encrypted by the CTR cipher without padding. The AES CTR cipher
+must be implemented by a cryptographic provider according to the NIST SP 800-38A specification. 
+
+Unique nonces are generated for each ciphertext, with the nonce length of 12 bytes. All 12 
+bytes are constructed with an RBG option 1 as described for the AES_GCM_V1 algorithm. The data 
+module encryption and decryption is performed with a 16 byte IV comprised of the nonce and a 4-byte 
+initial counter field. The first 31 bits of the initial counter field are set to 0, the last bit 
+is set to 1. 
+
+### 4.3 Key metadata
+A wide variety of services and tools for management of encryption keys exist in the 
+industry today. Public clouds offer different key management services (KMS), and 
+organizational IT systems either build proprietary key managers in-house or adopt open source 
+tools for on-premises deployment. Besides the diversity of management tools, there are many 
+ways to generate and handle the keys themselves (generate Data keys inside KMS – or locally 
+upon data encryption; use Data keys only, or use Master keys to encrypt/wrap the Data keys; 
+store the wrapped key material inside the data file, or at a separate location; etc). There 
+is also a large variety of authorization and certification methods, required to control the 
+access to encryption keys.
+
+Parquet is not limited to a single KMS, key generation/wrapping method, or authorization service. 
+Instead, Parquet provides a developer with a simple interface that can be utilized for implementation 
+of any key management scheme. For each column or footer key, a file writing code can pass an 
+arbitrary `key_metadata` byte array that will be stored in the file. A file reading code can pass a 
+key retrieval callback, that will invoked on the key_metadata byte array in order to recover the
+encryption (or signature) key. For example, the key_metadata can keep a serialized
+
+   * String ID of a Data encryption key (for direct key retrieval from a KMS).
+   * Wrapped Data key, and string ID of a Master key. The Data key is generated randomly and 
+   wrapped either remotely in a KMS, or locally after retrieving the Master key from KMS. Wrapping 
+   format depends on the KMS and can be a JSON string, or a base64 encoded byte array in a case of 
+   local wrapping.
+   * Short ID (counter) of a key inside the Parquet data file. The Data key is wrapped with a 
```

Comment:
I think the first two examples and the last one are sufficient. This one and the next aren't particularly valuable. In fact, the last one could be dropped as well for brevity.

Reply:
Unlike the second example, this one (third) allows for master key rotation without modifying the data file (footer), so might end up being the most useful among the three. Dropping the last two examples, and adding the key rotation point to the 2nd and 3rd.